lib/import.js

```diff
@@ -0,0 +1,227 @@
+
+      const data = await readFile(path);
+
+      return {
+        data,
+        updatedMap: map
+      };
+
+    } else {
+      self.apos.util.error(path + ': no path provided, cannot continue');
+      return {
+        data: null,
+        updatedMap: map
+      };
+    }
+  }
+
+  async function parseMap(xml = '', map = {}) {
+    const parseString = require('util').promisify(xml2js.parseString);
+
+    const svgs = [];
+    const result = await parseString(xml);
+
+    let symbols = findInObj(result, 'symbol');
+
+    if (!symbols.length) {
+      self.apos.util.error('Could not find an array of <symbol> elements in map ' + map.label);
+
+      throw self.apos.error('invalid');
+    }
+
+    if (symbols[0] && symbols[0].symbol) {
+      symbols = symbols[0].symbol;
+    } else {
+      self.apos.util.error('Error occurred parsing array of symbols in map ' + map.label);
+
+      throw self.apos.error('error');
+    }
+
+    symbols.forEach(function (symbol) {
+      if (symbol.$.id) {
+        svgs.push({
+          symbol: symbol.$,
+          file: map.finalFile || map.file,
+          map: map.name
+        });
+      } else {
+        self.apos.util.error('SVG is malformed or has no ID property');
+
+        throw self.apos.error('invalid');
+      }
+    });
+
+    return svgs;
+
+  }
+
+  async function evaluateForUpsert(svgs) {
+    const req = self.apos.task.getReq();
+
+    for (const svg of svgs) {
+
+      const docs = await self.find(req, {
+        svgId: svg.symbol.id
+      }, {}).toArray();
+
+      if (docs.length) {
+        // i have a doc, update it
+        await updatePiece(docs[0], svg);
+      } else {
+        // i don't have a doc, insert it
+        await insertPiece(svg);
+      }
+    }
+  }
+
+  async function insertPiece(svg) {
+    const req = self.apos.task.getReq();
+    const piece = self.newInstance();
+
+    if (svg.symbol.title) {
+      piece.title = self.apos.launder.string(svg.symbol.title);
+    } else {
+      piece.title = self.apos.launder.string(svg.symbol.id);
+    }
+
+    piece.svgId = svg.symbol.id;
+    piece.file = svg.file;
+    piece.map = svg.map;
+
+    await self.insert(req, piece, {
+      permissions: false
+    });
+  }
+
+  async function updatePiece(doc, svg, callback) {
+    const updateFields = {};
+
+    if (svg.symbol.title) {
+      updateFields.title = self.apos.launder.string(svg.symbol.title);
+    } else {
+      updateFields.title = self.apos.launder.string(svg.symbol.id);
+    }
+
+    updateFields.file = svg.file;
+    updateFields.map = svg.map;
+
+    await self.apos.doc.db.updateOne({
+      _id: doc._id
+    }, {
+      $set: updateFields
+    });
+  }
+
+  function fileExists(path) {
+    if (fs.existsSync(path)) {
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  function findInObj(obj, key) {
+
+    if (_.has(obj, key)) {
+      return [ obj ];
+    }
+    // TODO: replace lodash? or use flatMap?
```

Comment:
Supported back to node 12, so fine to use flat or flatMap for A3. Also fine to use lodash though.